### PR TITLE
Add `groupInto` to `Array` and `Seq`.

### PR DESCRIPTION
**Pull request by He-Pin**
For keeping their original ordering.

refs:https://github.com/scala/bug/issues/11276

cc @hrhino

### CODE REVIEW

**Comment by Jasper-M** [comment]
It's not 100% clear to me what it means for `groupby` to "retain the original ordering"? As far as I can see the groups already keep their ordering. And the keys are not actually *in* the original collection.

**Comment by He-Pin** [comment]
@Jasper-M The `groupBy` return a `Map` but not a `ListMap` or `VectorMap`, when you iterating through the returned map, the ordering of the grouped key not alias with their original ordering. Take the example from the PR, you may first iteration the batch of `2` and then `1`.

This seems not to be a problem when the grouped keys are less than 4, because of `Map1...4` thing, But will cause a problem with more grouped keys.

refs：https://github.com/scala/scala-async/pull/203
refs: https://github.com/scala/scala/pull/7429

**Comment by Jasper-M** [comment]
@hepin1989 But that what is tested in the PR is already what `groupBy` does:

```
scala> Seq.range(0, 100).groupBy(a => a % 10).values.forall(group => group.sorted == group)
res47: Boolean = true
```

**Comment by Jasper-M** [comment]
I think I have 2 issues here:
1. IMHO you can't really talk about original iteration order of the keys since the keys didn't exist before calling `[stable]GroupBy`. So there should be a better way to describe the difference between these methods. 
2. The tests are not actually testing the new behavior.

Also not really sure if it's worth adding extra methods for this.

**Comment by He-Pin** [comment]
@Jasper-M I will add more keys,it currenlty only have 3 group is
image
with pr
image

**Comment by He-Pin** [comment]
@Jasper-M I could update the test, yes, it was not good. For the description which could be refined too.
But I really want to get this PR to get merged, because not much surprising then.

**Comment by Jasper-M** [comment]
I might have been nitpicking a bit too much... If other people think it's useful to have these methods next to the regular `groupBy` that's fine by me...

**Comment by He-Pin** [comment]
@Jasper-M You are welcome, we are a community, the more review the far better right?
I will update the PR with the tests your suggested when I got home, back to the method itself, I just can't come up with another name.

File: test/junit/scala/collection/SeqTest.scala [hunk 1]
```diff
@@ -90,4 +90,36 @@ class SeqTest {
     assert(s.lengthIs > 2)
     assert(s.lengthIs != 2)
   }
+
+  @Test
+  def stableGroupBy(): Unit = {
+    val keys = Seq.range(0, 100).stableGroupBy(a => a % 10).keys.toArray
```
Comment (He-Pin) [inline test/junit/scala/collection/SeqTest.scala]: @Jasper-M I have updated the test.

File: src/library/scala/collection/Seq.scala [hunk 2]
```diff
@@ -277,6 +275,73 @@ trait SeqOps[+A, +CC[_], +C] extends Any
...
+  def stableGroupBy[K](f: A => K): immutable.SeqMap[K, C] = {
+    val m = mutable.SeqMap.empty[K, Builder[A, C]]
+    val it = iterator
+    while (it.hasNext) {
+      val elem = it.next()
+      val key = f(elem)
+      val bldr = m.getOrElseUpdate(key, newSpecificBuilder)
+      bldr += elem
+    }
+    var result = immutable.SeqMap.empty[K, C]
```
Comment (jasonqu) [inline src/library/scala/collection/Seq.scala]: why not use `m.view.mapValues(_.result()).toSeqMap` as you did in `ArrayOps.scala` :)

File: src/library/scala/collection/Seq.scala [hunk 3]
```diff
@@ -277,6 +275,73 @@ trait SeqOps[+A, +CC[_], +C] extends Any
...
+    */
+  def stableGroupMap[K, B](key: A => K)(f: A => B): immutable.SeqMap[K, CC[B]] = {
+    val m = mutable.SeqMap.empty[K, Builder[B, CC[B]]]
+    for (elem <- this) {
+      val k = key(elem)
+      val bldr = m.getOrElseUpdate(k, iterableFactory.newBuilder[B])
+      bldr += f(elem)
+    }
+    var result = immutable.SeqMap.empty[K, CC[B]]
+    m.foreach { case (k, v) =>
```
Comment (jasonqu) [inline src/library/scala/collection/Seq.scala]: same issue, just curious, why not use `m.view.mapValues(_.result()).toSeqMap`

File: src/library/scala/collection/Seq.scala (hunk 2 above)
Comment (He-Pin) [reply src/library/scala/collection/Seq.scala]: @jasonqu These methods are copied from `groupBy` in `Iterable` , and yes I could do that.

**Review by jasonqu — APPROVED** [review]
cool

**Comment by He-Pin** [comment]
I have rebased.

**Comment by He-Pin** [comment]
rebased.

**Comment by Ichoran** [comment]
@hepin1989 - Upon review, I don't think this is a sufficiently general-purpose method.  All that's happening here is that the collection type is being specialized.  But why does SeqMap warrant special attention?  What if you want to group into a mutable map so you can change it?  What if you have int keys and want higher efficiency?  What if you want a concurrent map?

A method on the companions of all maps called `groupFrom` or somesuch would be a great improvement.  (Alternatively, a `groupInto` method that also takes as a parameter the companion of the map you want to build, which would use the same mechanics as the `to()` method.)  I'll leave this PR open in case you want to rework this one in that more general fashion.

You could also close this and open a new PR.  (But soon!  RC1 is almost upon us!)

**Comment by He-Pin** [comment]
How about open up a RC2 and rescheduling it to there？
From a practical point, I this this one is ok, but yes could be implemented on top of something more generally,then need to touch both methods.

**Comment by He-Pin** [comment]
If I find sometime, I can do it tomorrow 😂

**Comment by Ichoran** [comment]
It's better to delay RC1 than have an RC2 with new features.  We have at least a couple of weeks still to try to get this in in a generalized form.

**Comment by He-Pin** [comment]
You mean something like this? @Ichoran 
```scala
  def groupBy[K](f: A => K): immutable.Map[K, C] = {
    groupInto(f,iterableFactory,HashMap)
  }

  def groupInto[K, C1[_], MC[_, _]](f: A => K, 
                                    valueFactory: IterableFactory[C1], 
                                    mapFactory: MapFactory[MC]): MC[K, C1] = {
    mapFactory.groupFrom(this, f, valueFactory)
  }
```

**Comment by Ichoran** [comment]
@hepin - I was thinking `groupInto[K, MC[_, _]](mapFactory: MapFactory[MC])(f: A => K): MC[K, C]`.  Generalizing the target collection as well is a bit much, I think, for general use.

**Comment by julienrf** [comment]
FWIW, we do already have a general form: `groupMapReduce`.

**Comment by Ichoran** [comment]
`groupMapReduce` is a different axis of generalization than `groupInto`; the latter lets you pick the type of the target map, while the former lets you pick which operations to perform on the grouped elements in your standard immutable map.

File: src/library/scala/collection/Iterable.scala [hunk 4]
```diff
@@ -537,21 +537,11 @@ trait IterableOps[+A, +CC[_], +C] extends Any with IterableOnce[A] with Iterable
...
-    while (mapIt.hasNext) {
-      val (k, v) = mapIt.next()
-      result = result.updated(k, v.result())
-    }
-    result
+    groupInto(immutable.Map)(f)
+  }
+
+  def groupInto[K, MC[_, _]](mapFactory: MapFactory[MC])(f: A => K): MC[K, C @uncheckedVariance] = {
+    mapFactory.groupFrom(this, f, self.newSpecificBuilder)
```
Comment (He-Pin) [inline src/library/scala/collection/Iterable.scala:544]: Add groupInto

File: src/library/scala/collection/ArrayOps.scala [hunk 5]
```diff
@@ -1349,17 +1349,11 @@ final class ArrayOps[A](val xs: Array[A]) extends AnyVal {
...
-      val bldr = m.getOrElseUpdate(key, ArrayBuilder.make[A])
-      bldr += elem
-      i += 1
-    }
-    m.view.mapValues(_.result()).toMap
+    groupInto(immutable.Map)(f)
+  }
+
+  def groupInto[K, MC[_, _]](mapFactory: MapFactory[MC])(f: A => K): MC[K, Array[A]] = {
+    mapFactory.groupFrom(view, f, ArrayBuilder.make[A])
```
Comment (He-Pin) [inline src/library/scala/collection/ArrayOps.scala:1356]: Add groupInto

**Comment by He-Pin** [comment]
@Ichoran Thanks for the guidance, and very sorry for the delay, I am currently busy on KPI things.
I updated the PR but without document yet. Once you think the code itself is fine, I can add the document later then.

**Comment by Ichoran** [comment]
The tests aren't passing, and you've changed and duplicated a _lot_ of code.  I'm sorry you put so much effort into this, but can you use a more simplistic strategy?  Check whether the mapFactory is giving you mutable or immutable maps; if mutable, build right into it using the existing algorithm.  If immutable, get a mutable map instead, build into it using the existing algorithm, and then convert using a builder.

So basically the existing code should be unchanged except the map might be a different mutable map; and instead of `toMap` at the end you might just return the map you have (if you got it from the factory) or use the factory to `from` it into the type you asked for.

**Comment by szeiger** [comment]
It's getting quite late in the RC cycle. This looks safe to reschedule to 2.14 unless we get to a mergeable version within the next 2 weeks.

**Comment by adriaanm** [comment]
Rescheduling to 2.14 based on Stefan's comment -- if this becomes mergeable this week, please ping me and I'll move it back.

**Comment by dwijnand** [comment]
This is about a week shy from being a year since the last commit.  Should we close this one?  It's WIP and got a bunch of conflicts.  Of course closing it doesn't make the code irrecoverable and, further, it's nicely linking to scala/bug#11276, which would stay open.
